tests/integration/test_increment_graph_window.py

```diff
@@ -94,6 +94,24 @@ def add_task(schd, task):
     schd.data_store_mgr.add_pool_node(task, IntegerPoint('1'))
 
 
+def get_graph_walk_cache(schd):
+    """Return the head task names of cached graph walks."""
+    # prune graph walk cache
+    schd.data_store_mgr.prune_data_store()
+    # fetch the cached walks
+    n_window_node_walks = sorted(
+        Tokens(task_id)['task']
+        for task_id in schd.data_store_mgr.n_window_node_walks
+    )
+    n_window_completed_walks = sorted(
+        Tokens(task_id)['task']
+        for task_id in schd.data_store_mgr.n_window_completed_walks
+    )
+    # TODO: these two dicts appear to always contain the same things?
+    assert n_window_node_walks == n_window_completed_walks
+    return n_window_completed_walks
```

Comment:
I spotted that `n_window_node_walks` appears to contain the same keys as `n_window_completed_walks` which suggests that we only need one of them.

Could we consolidate this down to one dictionary?

Reply:
Don't think so... `n_window_completed_walks` is a set that contains active/ex-active ids of walks that have been completed... And in larger windows there can be incomplete walks that are a part of `n_window_node_walks` (which is a dictionary)..

Reply:
So is  `n_window_completed_walks` for completed walks and `n_window_node_walks` for incomplete? If so, why are the same keys in both?

Reply:
Because I lookup `n_window_completed_walks` (which is a **set**) with the keys from `n_window_node_walks` (which is a **dictionary**) that may include keys that are not in `n_window_completed_walks` ..

It's used to re-walk completed walks if they become active again (which would still use info from walks around them)...  

Looking at the code again, I believe it's for regenerating the window/data-store about past task still in `n_window_node_walks` that gets run again ... 

Although I could test whether it's still needed, I'm struggling to understand the issue here, one is a set to lookup particular types of walks (by their ID), the other is a dictionary of the walk information with the ID as key.

Reply:
> Although I could test whether it's still needed

Actually, had a think, and it is needed; a node on the boundary of the window will still have it's walk (which it should, for building other walks), but nodes in it's walk will not be in the window/data-store.. The data-store is generated along the walk (where it hasn't been walked), so if you rerun a node on the boundary you need to re-walk the parts outside of the window..

Reply:
> So is `n_window_completed_walks` for completed walks and `n_window_node_walks` for incomplete? If so, why are the same keys in both?

So in short;
- they are not the same (one is a set the other a dictionary)
- the set of keys and the keys in the dictionary can only be the same if;
  - they are empty (no task have been populated)
  - all tasks in the window are active/ex-active  (i.e. n=0 window or something else wierd like no future tasks (at the FCP))

Reply:
Thanks for the explanation. No issue here, I'm just trying to understand how the bits fit together.